tests/test_domain_py.py

```diff
@@ -44,3 +45,52 @@ def test_function_signatures():
 
     rv = parse('func(a=[][, b=None])')
     assert text_type(rv) == u'a=[], [b=None]'
+
+
+@pytest.mark.sphinx(testroot='domain-py')
+def test_build_domain_py_xrefs_resolve_correctly(app, status, warning):
+    from sphinx.domains.python import PythonDomain
+
+    calls = {}
+
+    def wrapped_find_obj(fn):
+        def wrapped(*args):
+            ret = fn(*args)
+            # args = [domain, env, ??, env object, role text, role type, order]
+            calls[args[3:-1]] = ret
+            return ret
+        return wrapped
+
+    PythonDomain.find_obj = wrapped_find_obj(PythonDomain.find_obj)
+
+    app.builder.build_all()
+
+    calls_expected = {
+        (None, u'TopLevel', u'class'): [
+            (u'TopLevel', (u'roles', u'class'))],
+        (None, u'top_level', u'meth'): [
+            (u'top_level', (u'roles', u'method'))],
+        (u'NestedParentA', u'child_1', u'meth'): [
+            (u'NestedParentA.child_1', (u'roles', u'method'))],
+        (u'NestedParentA', u'NestedChildA.subchild_2', u'meth'): [
+            (u'NestedParentA.NestedChildA.subchild_2', (u'roles', u'method'))],
+        (u'NestedParentA', u'child_2', u'meth'): [
+            (u'child_2', (u'roles', u'method'))],
+        (u'NestedParentA', u'any_child', None): [
+            (u'NestedParentA.any_child', (u'roles', u'method'))],
+        (u'NestedParentA', u'NestedChildA', u'class'): [
+            (u'NestedParentA.NestedChildA', (u'roles', u'class'))],
+        (u'NestedParentA.NestedChildA', u'subchild_2', u'meth'): [
+            (u'NestedParentA.NestedChildA.subchild_2', (u'roles', u'method'))],
+        (u'NestedParentA.NestedChildA', u'NestedParentA.child_1', u'meth'): [
+            (u'NestedParentA.child_1', (u'roles', u'method'))],
+        (None, u'NestedChildA.subchild_1', u'meth'): [],
```

Comment:
For instance, line 87 here is buggy behavior that I've patched with #3465. The arguments in to `find_obj` are not expected, and the return from `find_obj` is empty. Instead of:

```python
(None, u'NestedChildA.subchild_1', u'meth'): [],
```

This should instead be:

```python
(u'NestedParentA.NestedChildA', u'subchild_1', u'meth'): [
    (u'NestedParentA.NestedChildA.subchild_1', (u'roles', u'method'))
],
```